token/program-2022/src/offchain.rs

```diff
@@ -46,14 +61,1005 @@ where
+            Err(TokenError::InvalidInstruction)?;
+        }
+
+        let mut execute_ix = spl_transfer_hook_interface::instruction::execute(
+            &program_id,
+            &instruction.accounts[0].pubkey,
+            &instruction.accounts[1].pubkey,
+            &instruction.accounts[2].pubkey,
+            &instruction.accounts[3].pubkey,
+            &get_extra_account_metas_address(mint_address, &program_id),
+            amount,
+        );
+
         resolve_extra_account_metas(
-            instruction,
+            &mut execute_ix,
             fetch_account_data_fn,
             mint_address,
             &program_id,
         )
         .await?;
+
+        instruction
+            .accounts
+            .extend_from_slice(&execute_ix.accounts[5..]);
     }
     Ok(())
 }
+
+/// Offchain helper to create a `TransferChecked` instruction with all
+/// additional required account metas.
+///
+/// To be client-agnostic and to avoid pulling in the full solana-sdk, this
+/// simply takes a function that will return its data as `Future<Vec<u8>>` for
+/// the given address. Can be called in the following way:
+///
+/// ```rust,ignore
+/// use futures_util::TryFutureExt;
+/// use solana_client::nonblocking::rpc_client::RpcClient;
+/// use solana_program::pubkey::Pubkey;
+///
+/// let mint = Pubkey::new_unique();
+/// let client = RpcClient::new_mock("succeeds".to_string());
+/// let mut account_metas = vec![];
+///
+/// get_extra_transfer_account_metas(
+///     &mut account_metas,
+///     |address| self.client.get_account(&address).map_ok(|opt| opt.map(|acc| acc.data)),
+///     &mint,
+/// ).await?;
+/// ```
+/// Note that this offchain helper will build a new `Execute` instruction,
+/// resolve the extra account metas, and then add them to the transfer
+/// instruction. This is because the extra account metas are configured
+/// specifically for the `Execute` instruction, which requires five accounts
+/// (source, mint, destination, authority, and validation state), wheras the
+/// transfer instruction only requires four (source, mint, destination, and
+/// authority) in addition to `n` number of multisig authorities.
+#[allow(clippy::too_many_arguments)]
+pub async fn create_transfer_instruction_with_extra_metas<F, Fut>(
+    token_program_id: &Pubkey,
+    source_pubkey: &Pubkey,
+    mint_pubkey: &Pubkey,
+    destination_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
+    signer_pubkeys: &[&Pubkey],
+    amount: u64,
+    decimals: u8,
+    fetch_account_data_fn: F,
+) -> Result<Instruction, AccountFetchError>
+where
+    F: Fn(Pubkey) -> Fut,
+    Fut: Future<Output = AccountDataResult>,
+{
+    let mut transfer_ix = crate::instruction::transfer_checked(
+        token_program_id,
+        source_pubkey,
+        mint_pubkey,
+        destination_pubkey,
+        authority_pubkey,
+        signer_pubkeys,
+        amount,
+        decimals,
+    )?;
+
+    if token_program_id == &crate::id() {
```

Comment:
What's this check for?

Reply:
Couldn't this be legacy token program? Are OK with just letting this gracefully return based on trying to find the transfer hook extension?

Reply:
legacy mints can be properly unpacked with `StateWithExtensions`, so i don't see a specific reason for this check